internal/docker/container.go

```diff
@@ -59,89 +78,124 @@ func (r *ContainerRunner) pullImage(img string) error {
 		}
-		confd.Image = img
+		docker.Image = img
 	}
 
-	if err := r.pullImage(confd.Image); err != nil {
+	if err := r.pullImage(docker.Image); err != nil {
 		return err
 	}
+	return nil
+}
 
-	container, err := r.docker.StartContainer(r.Ctx, files, confd)
+func (r *ContainerRunner) startContainer(options containerStartOptions) (string, error) {
+	container, err := r.docker.StartContainer(r.Ctx, options)
 	if err != nil {
-		return err
+		return "", err
 	}
-	r.containerID = container.ID
+	containerID := container.ID
 
-	pDir, err := r.docker.ProjectDir(r.Ctx, confd.Image)
+	pDir, err := r.docker.ProjectDir(r.Ctx, options.Docker.Image)
 	if err != nil {
-		return err
+		return "", err
+	}
+
+	r.containerConfig.jobDetailsFilePath, err = r.docker.JobDetailsURLFile(r.Ctx, options.Docker.Image)
+	if err != nil {
+		return "", err
 	}
 
 	tmpDir, err := ioutil.TempDir("", "saucectl")
 	if err != nil {
-		return err
+		return "", err
 	}
 	defer os.RemoveAll(tmpDir)
 
 	rcPath := filepath.Join(tmpDir, SauceRunnerConfigFile)
-	if err := jsonio.WriteFile(rcPath, project); err != nil {
-		return err
+	if err := jsonio.WriteFile(rcPath, options.Project); err != nil {
+		return "", err
 	}
 
-	if err := r.docker.CopyToContainer(r.Ctx, r.containerID, rcPath, pDir); err != nil {
-		return err
+	if err := r.docker.CopyToContainer(r.Ctx, containerID, rcPath, pDir); err != nil {
+		return "", err
 	}
 	r.containerConfig.sauceRunnerConfigPath = path.Join(pDir, SauceRunnerConfigFile)
 
 	// running pre-exec tasks
-	err = r.beforeExec(beforeExec)
+	err = r.beforeExec(containerID, options.SuiteName, options.BeforeExec)
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	return nil
+	return container.ID, nil
 }
 
-func (r *ContainerRunner) run(cmd []string, env map[string]string) error {
+func (r *ContainerRunner) run(containerID, suiteName string, cmd []string, env map[string]string) (string, string, string, bool, error) {
 	defer func() {
-		log.Info().Msg("Tearing down environment")
-		if err := r.docker.Teardown(r.Ctx, r.containerID); err != nil {
+		log.Info().Msgf("%s: Tearing down environment", suiteName)
+		if err := r.docker.Teardown(r.Ctx, containerID); err != nil {
 			if !r.docker.IsErrNotFound(err) {
-				log.Error().Err(err).Msg("Failed to tear down environment")
+				log.Error().Err(err).Msgf("%s: Failed to tear down environment", suiteName)
 			}
 		}
 	}()
 
-	exitCode, err := r.docker.ExecuteAttach(r.Ctx, r.containerID, r.Cli, cmd, env)
-	log.Info().
-		Int("ExitCode", exitCode).
-		Msg("Command Finished")
+	exitCode, output, err := r.docker.ExecuteAttach(r.Ctx, containerID, cmd, env)
 
 	if err != nil {
-		return err
+		return "", "", "", false, err
 	}
+
+	passed := true
 	if exitCode != 0 {
-		return fmt.Errorf("exitCode is %d", exitCode)
+		err = fmt.Errorf("exitCode is %d", exitCode)
+		passed = false
 	}
-	return nil
+
+	jobDetailsURL, err := r.readTestURL(containerID)
```

Comment:
Hm, this would now override the error that was set via `err = fmt.Errorf("exitCode is %d", exitCode)` 🤔

Reply:
@alexplischke I removed the previous `err = fmt.Errorf("exitCode is %d", exitCode)` statement, as the information is already carried by the `passed` value.

So there is no more overwriting of values, `err` contains purely technicals errors, `passed` contains the status of the test suite.